src/location.h

```diff
@@ -0,0 +1,248 @@
+  enum Mask : uint8_t {
+    NOT_INITIALIZED = 0,
+    HAS_LATLON = 1 << 0,
+    HAS_ALTITUDE = 1 << 1,
+    HAS_BEARING = 1 << 2,
+    HAS_SPEED = 1 << 3,
+    HAS_SOURCE = 1 << 4
+  } valid_values_mask_ = NOT_INITIALIZED;
+
+ public:
+  // Milliseconds from January 1, 1970.
+  uint64_t timestamp_ms_;
+  double latitude_deg_;
+  double longitude_deg_;
+  double horizontal_accuracy_m_;
+  double altitude_m_;
+  double vertical_accuracy_m_;
+  // Positive degrees from the true North.
+  double bearing_deg_;
+  // Metres per second.
+  double speed_mps_;
+
+  // Compacts location into the byte representation.
+  // TODO(AlexZ): We don't care about endiannes for now.
+  std::string Encode() const {
+    std::string s;
+    s.push_back(valid_values_mask_);
+    if (valid_values_mask_ == NOT_INITIALIZED) {
+      return s;
+    }
+    if (valid_values_mask_ & HAS_LATLON) {
+      static_assert(sizeof(timestamp_ms_) == 8, "We are cutting off timestamp to 6 bytes instead of 8.");
+      s.append(reinterpret_cast<const char *>(&timestamp_ms_), sizeof(timestamp_ms_) - 2);
+      const int32_t lat10mil = latitude_deg_ * 10000000.;
+      s.append(reinterpret_cast<const char *>(&lat10mil), sizeof(lat10mil));
+      const int32_t lon10mil = longitude_deg_ * 10000000.;
+      s.append(reinterpret_cast<const char *>(&lon10mil), sizeof(lon10mil));
+      const uint32_t horizontal_accuracy_cm = horizontal_accuracy_m_ * 100.;
+      s.append(reinterpret_cast<const char *>(&horizontal_accuracy_cm), sizeof(horizontal_accuracy_cm));
+      if (valid_values_mask_ & HAS_SOURCE) {
+        s.push_back(source_);
+      }
+    }
+    if (valid_values_mask_ & HAS_ALTITUDE) {
+      const int32_t altitude_cm = altitude_m_ * 100.;
+      s.append(reinterpret_cast<const char *>(&altitude_cm), sizeof(altitude_cm));
+      const uint16_t vertical_accuracy_cm = vertical_accuracy_m_ * 100.;
+      s.append(reinterpret_cast<const char *>(&vertical_accuracy_cm), sizeof(vertical_accuracy_cm));
+    }
+    if (valid_values_mask_ & HAS_BEARING) {
+      const uint32_t bearing10mil = bearing_deg_ * 10000000.;
+      s.append(reinterpret_cast<const char *>(&bearing10mil), sizeof(bearing10mil));
+    }
+    if (valid_values_mask_ & HAS_SPEED) {
+      const uint16_t speedx100mps = speed_mps_ * 100.;
+      s.append(reinterpret_cast<const char *>(&speedx100mps), sizeof(speedx100mps));
+    }
+    return s;
+  }
+
+  // Initializes location from serialized byte array created by ToString() method.
+  // TODO(AlexZ): we don't care about endianness right now.
+  explicit Location(const std::string &encoded) { Decode(encoded); }
+
+  void Decode(const std::string &encoded) {
+    if (encoded.empty()) {
+      return;
+    }
+    std::string::size_type i = 0;
+    valid_values_mask_ = static_cast<Mask>(encoded[i++]);
+    if (valid_values_mask_ == NOT_INITIALIZED) return;
+    if (valid_values_mask_ & HAS_LATLON) {
+      timestamp_ms_ = *reinterpret_cast<const uint32_t *>(&encoded[i]) |
```

Comment:
Same here: You could use a static function. Also, keeping a ```const uint8_t* ptr``` would make your life easier by eliminating ```i```.